java/ql/src/semmle/code/java/security/XPath.qll

```diff
@@ -0,0 +1,65 @@
+/** Provides classes to reason about XPath vulnerabilities. */
+
+import java
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.dataflow.TaintTracking
+
+/**
+ * An abstract type representing a call to interpret XPath expressions.
+ */
+class XPathSink extends MethodAccess {
+  /**
+   * Gets the argument representing the XPath expressions to be evaluated.
+   */
+  abstract Expr getSink();
+}
+
+/** The class `javax.xml.xpath.XPath` */
+class XPath extends RefType {
+  XPath() { this.hasQualifiedName("javax.xml.xpath", "XPath") }
+}
+
+/** A call to `XPath.evaluate` or `XPath.compile` */
+class XPathEvaluateOrCompile extends XPathSink {
+  XPathEvaluateOrCompile() {
+    exists(Method m | this.getMethod() = m and m.getDeclaringType() instanceof XPath |
+      m.hasName(["evaluate", "compile"])
+    )
+  }
+
+  override Expr getSink() { result = this.getArgument(0) }
+}
+
+/** Any class extending or implementing `org.dom4j.Node` */
+class Dom4JNode extends RefType {
+  Dom4JNode() {
+    exists(Interface node | node.hasQualifiedName("org.dom4j", "Node") |
+      this.extendsOrImplements*(node)
+    )
+  }
+}
+
+/** A call to `Node.selectNodes` or `Node.selectSingleNode` */
+class NodeSelectNodes extends XPathSink {
+  NodeSelectNodes() {
+    exists(Method m | this.getMethod() = m and m.getDeclaringType() instanceof Dom4JNode |
+      m.hasName(["selectNodes", "selectSingleNode"])
```

Comment:
Might be good to consider the following methods:
- [`matches`](https://javadoc.io/static/org.dom4j/dom4j/2.1.3/org/dom4j/Node.html#matches(java.lang.String))
- [`numberValueOf`](https://javadoc.io/static/org.dom4j/dom4j/2.1.3/org/dom4j/Node.html#numberValueOf(java.lang.String))
- [`selectObject`](https://javadoc.io/static/org.dom4j/dom4j/2.1.3/org/dom4j/Node.html#selectObject(java.lang.String))
- [`valueOf`](https://javadoc.io/static/org.dom4j/dom4j/2.1.3/org/dom4j/Node.html#valueOf(java.lang.String))

Additionally [`createXPath`](https://javadoc.io/static/org.dom4j/dom4j/2.1.3/org/dom4j/Node.html#createXPath(java.lang.String)) is probably relevant as well since the created XPath is most likely evaluated afterwards.

Reply:
All those seem valid sinks, thanks! Added in ab543de